Amazon.QLDB.Driver/result/generic/AsyncResult.cs

```diff
@@ -0,0 +1,171 @@
+            this.ionEnumerator = new IonAsyncEnumerator(session, txnId, statementResult, cancellationToken, query);
+        }
+
+        /// <inheritdoc/>
+        public IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+        {
+            if (this.isRetrieved)
+            {
+                throw new InvalidOperationException();
+            }
+
+            this.isRetrieved = true;
+            if (cancellationToken != default)
+            {
+                this.ionEnumerator.CancellationToken = cancellationToken;
+            }
+
+            return this.ionEnumerator;
+        }
+
+        /// <summary>
+        /// Gets the current query statistics for the number of read IO requests. The statistics are stateful.
+        /// </summary>
+        ///
+        /// <returns>The current IOUsage statistics.</returns>
+        public IOUsage? GetConsumedIOs()
+        {
+            return this.ionEnumerator.GetConsumedIOs();
+        }
+
+        /// <summary>
+        /// Gets the current query statistics for server-side processing time. The statistics are stateful.
+        /// </summary>
+        ///
+        /// <returns>The current TimingInformation statistics.</returns>
+        public TimingInformation? GetTimingInformation()
+        {
+            return this.ionEnumerator.GetTimingInformation();
+        }
+
+        /// <summary>
+        /// Object which allows for asynchronous iteration over the individual Ion values that make up the whole result
+        /// of a statement execution against QLDB.
+        /// </summary>
+        private class IonAsyncEnumerator : BaseIonEnumerator, IAsyncEnumerator<T>
+        {
+            internal CancellationToken CancellationToken;
+            private readonly IQuery<T> query;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="IonAsyncEnumerator"/> class.
+            /// </summary>
+            ///
+            /// <param name="session">The parent session that represents the communication channel to QLDB.</param>
+            /// <param name="txnId">The unique ID of the transaction.</param>
+            /// <param name="statementResult">The result of the statement execution.</param>
+            /// <param name="cancellationToken">The CancellationToken to use for this AsyncResult.</param>
+            internal IonAsyncEnumerator(
+                Session session,
+                string txnId,
+                QLDBSession.Model.ExecuteStatementResult statementResult,
+                CancellationToken cancellationToken, 
+                IQuery<T> query)
+                : base(session, txnId, statementResult)
+            {
+                this.CancellationToken = cancellationToken;
+                this.query = query;
+            }
+
+            new public T Current
+            {
+                get
+                {
+                    return query.Deserialize(this.currentEnumerator.Current);
```

Comment:
What if `query.serializer` is null? The default behavior in the driver is to set serializer to null

Reply:
Added a null check in [QueryData constructor.](https://github.com/awslabs/amazon-qldb-driver-dotnet/blob/object-serialization-idda-jun-4-2021/Amazon.QLDB.Driver/query/Query.cs#L52-L55)